src/puppetlabs/puppetdb/utils.clj

```diff
@@ -343,6 +345,29 @@
     {:command command :version (Integer. version)
      :certname certname :payload (json/->RawJsonString payload)}))
 
+(defn utf8-bytes [s] (.getBytes ^String s "UTF-8"))
+
+(defn utf8-length
+  "Return the length in bytes of the given string when encoded in UTF-8"
+  [s]
+  (count (.getBytes ^String s "UTF-8")))
+
+(def ^Charset utf8-charset (Charset/forName "UTF-8"))
+
+(defn utf8-truncate
+  "Truncate the given string such that its UTF-8 representation is at most
+  `max-bytes` long. Note that the returned string may be empty."
+  [s max-bytes]
+  (let [utf8-bytebuff (ByteBuffer/allocate max-bytes)
+        chars-in (CharBuffer/wrap (chars (.toCharArray ^String s)))
+        encoder (.newEncoder utf8-charset)]
+    (.encode encoder chars-in utf8-bytebuff true)
+    (.flush encoder utf8-bytebuff)
+    (let [utf8-bytes (->> (.array utf8-bytebuff)
+                       (remove zero?)
+                       byte-array)]
```

Comment:
I'm not sure, but I think we might be able to replace this with something like (String. bytebuf 0 (.position bytebuf) utf8-charset), assuming the encoder puts position right at the end of the valid bytes.